core/src/util.rs

```diff
@@ -27,3 +27,187 @@ impl<T: Seq> SeqGenerator<T> {
         mem::replace(&mut self.0, value)
     }
 }
+
+pub trait PacketData: Copy {
+    fn write(self, vec: &mut Vec<u8>);
+
+    fn size_hint(&self) -> usize;
+}
+
+pub trait IntoPacketData<T = Self> {
```

Comment:
Why are both `PacketData` and `IntoPacketData` needed? Seems like `PacketData` would be enough.

Reply:
`IntoPacketData` is the type you use inside of the macro. `u32`, `[u8]`, `[u8; 20]` or `Proto`. It's just a placeholder to make it look nice, and can be freely chosen. `PacketData` represents the actual data.

One implementor of `PacketData` is `&[u8]`, but I thought it would look better to used `[u8]` instead of `&[u8]` in the macro, because the latter is "just a pointer".

Another implementor is `ProtoPacketData<T>` (or similar?), but even if it had a shorter name, it wouldn't look good in the macro because the generic parameter has to be mentioned.

So I decided to keep the type in the brackets completely separate of what actually happens to the passed data.

I still tried several times to unify these traits, but it didn't work out well.

Reply:
How about https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3ed7ed355bfaaad272e7969a8d51f5b2 ?

It relies on two key aspects. The first is the `impl PacketData for &'_ T` part, which is a [fairly common Rust idiom](https://doc.rust-lang.org/src/std/io/impls.rs.html#53-83) and the special `(proto) m` macro syntax.